### PR TITLE
fix(ui): stop rows showing through the pinned toolbar

### PR DESCRIPTION
**Pull request by ajay-fluxon**
## Summary

- **Sticky toolbar bleed-through.** `.test-cases-toolbar` is `position: sticky` but had no background of its own, so test-case rows scrolling underneath showed through it once it pinned to the top. It now gets an opaque background *only while actually stuck*, so it still blends with whatever background the host page provides in normal flow.
- **Stuck detection.** A zero-height sentinel sits just above the toolbar; an `IntersectionObserver` in `llm-test-runner.tsx` flips the new `isToolbarStuck` state when it leaves the viewport, which adds `.test-cases-toolbar--stuck` (white background, subtle border + shadow, 150ms transition). The observer accounts for the existing `stickyOffset` prop via `rootMargin` and is rebuilt when that prop changes, disconnected on `disconnectedCallback`, and skipped entirely when `IntersectionObserver` is undefined (the hydrate/SSR build and the jsdom test env both run this lifecycle in Node).
- **Primary button hover.** `.ui-button--primary` / `.ui-icon-button--primary` lightened on hover via `opacity: 0.9`, which blended the blue into the page background and read as washed out. They now darken using a dedicated `--primary-hover` token (`#0925b2` light, `#324dd9` dark).

## Test plan

- [x] `npm run lint` — 0 errors (7 pre-existing `no-explicit-any` warnings in untouched files)
- [x] `npx tsc --noEmit -p tsconfig.react.json` — clean
- [x] `npm test` — 226/226 tests pass (one flaky parallel-worker SIGSEGV, hits a different file each run and passes in isolation; unrelated to this change)
- [x] Manual: scroll the test-case list and confirm rows no longer show through the toolbar, that it's transparent in normal position and white only once pinned, and that Run / Run all darken rather than lighten on hover

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ajay-fluxon** [comment]
Loom Video: [Link](https://www.loom.com/share/9de313b0a96349d289c4e2a9c6eda5c4)

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 1]
```diff
@@ -133,10 +136,31 @@ export class LLMTestRunner {
...
   }
 
   componentDidLoad() {
     this.el.style.setProperty('--llmtr-sticky-top', `${this.stickyOffset ?? 0}px`);
+    this.setupToolbarStuckObserver();
+  }
+
+  disconnectedCallback() {
+    this.toolbarStuckObserver?.disconnect();
+  }
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx:161]: Good catch — fixed in ac8f839. Added a `connectedCallback` that calls `setupToolbarStuckObserver()`, so the observer is rebuilt on every reattach. The existing `!this.toolbarSentinelEl` guard makes the first pre-render call a no-op as you noted.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 2]
```diff
@@ -133,10 +136,31 @@ export class LLMTestRunner {
...
+  private setupToolbarStuckObserver() {
+    this.toolbarStuckObserver?.disconnect();
+    // Absent in Node — the hydrate/SSR build and the jsdom test env both run this.
+    if (!this.toolbarSentinelEl || typeof IntersectionObserver === 'undefined') {
+      return;
+    }
+    this.toolbarStuckObserver = new IntersectionObserver(
+      ([entry]) => {
+        this.isToolbarStuck = !entry.isIntersecting;
+      },
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx]: Agreed, this was a real bug — the toolbar rendered stuck in normal flow whenever the component sat below the fold. Fixed in ac8f839: now also requires `entry.boundingClientRect.top <= (entry.rootBounds?.top ?? 0)`, so only a sentinel at/above the adjusted top edge counts as stuck. Also bumped the sentinel to 1px, since zero-area targets are an IntersectionObserver edge case.

File: src/components/llm-test-runner/test-cases/test-cases-toolbar.css [hunk 3]
```diff
@@ -10,6 +12,13 @@
   position: sticky;
   top: var(--llmtr-sticky-top, 0px);
   z-index: var(--z-sticky);
+  transition: background-color 150ms ease, box-shadow 150ms ease;
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/test-cases/test-cases-toolbar.css]: Correct — fading up from transparent reintroduced the bleed-through for the length of the transition. Fixed in ac8f839: the transition is now `box-shadow` only, so the background flips to opaque immediately on pin.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 4]
```diff
@@ -133,10 +136,40 @@ export class LLMTestRunner {
...
+    }
+    this.toolbarStuckObserver = new IntersectionObserver(
+      ([entry]) => {
+        // The sentinel is also outside the root while the whole component sits
+        // below the fold, so isIntersecting alone would read as stuck there.
+        const rootTop = entry.rootBounds?.top ?? 0;
+        this.isToolbarStuck =
+          !entry.isIntersecting && entry.boundingClientRect.top <= rootTop;
+      },
+      { rootMargin: `-${(this.stickyOffset ?? 0) + 1}px 0px 0px 0px` },
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx]: Confirmed and fixed in e008e37 — and this was introduced by my own fix for the below-the-fold comment, since anchoring to `rootBounds` assumes the viewport is the scroll container.

Rather than resolving the scroll container, the check now compares the toolbar and sentinel rects directly: the sentinel abuts the toolbar until it pins, so a gap between them means stuck. Both rects share a coordinate space, so it holds wherever the scrolling happens, and the viewport assumption is gone.

Covered by a regression test (`is stuck when pinned inside a scroll container below the viewport top`) — verified it fails against the previous viewport-relative logic.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 5]
```diff
@@ -133,10 +136,40 @@ export class LLMTestRunner {
...
+
+  disconnectedCallback() {
+    this.toolbarStuckObserver?.disconnect();
+  }
+
+  private setupToolbarStuckObserver() {
+    this.toolbarStuckObserver?.disconnect();
+    // Absent in Node — the hydrate/SSR build and the jsdom test env both run this.
+    if (!this.toolbarSentinelEl || typeof IntersectionObserver === 'undefined') {
+      return;
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx]: Agreed — three correctness fixes in this block with no coverage was the right thing to flag. Added `src/components/llm-test-runner/tests/toolbarStuck.spec.tsx` in e008e37 with a mocked IntersectionObserver, covering:

- sentinel abutting the toolbar (not stuck) and below the fold (not stuck)
- crossed threshold (stuck) and pinned inside a scroll container below the viewport top (stuck)
- releasing the stuck state when the gap closes, and ignoring sub-pixel gaps
- observer recreation with an updated rootMargin when `stickyOffset` changes
- disconnect on detach and re-observe on reattach

9 tests; full suite is 238 passing. I also mutation-checked the scroll-container case to confirm it fails against the previous logic rather than passing trivially.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 6]
```diff
@@ -133,10 +136,39 @@ export class LLMTestRunner {
...
+    }
+    this.toolbarStuckObserver = new IntersectionObserver(
+      ([entry]) => {
+        // isIntersecting is also false below the fold, which isn't stuck.
+        const rootTop = entry.rootBounds?.top ?? 0;
+        this.isToolbarStuck =
+          !entry.isIntersecting && entry.boundingClientRect.top <= rootTop;
+      },
+      { rootMargin: `-${(this.stickyOffset ?? 0) + 1}px 0px 0px 0px` },
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx]: Correct about the off-by-one, and chasing it turned out to be worth more than the fix itself — it exposed that the whole observer approach was unsound here, so the `rootMargin` line this thread is about is gone as of 4cb6f9a.

Two problems, both from re-evaluating only at threshold crossings:

1. The gap-based verdict lags the real pin by a pixel or two. Scrolling slowly put the single callback inside that lag, so the toolbar pinned while still transparent and never corrected — rows visibly showed through it. Removing the `+ 1` narrowed the window but did not close it, since at the crossing the gap is still ~0.
2. A scroll that jumps straight into pin range (deep link, scroll restoration, `scrollIntoView` — which this component calls itself) never produces a crossing at all, so nothing fires.

Replaced with a rAF-throttled scroll listener in the capture phase, re-evaluating every frame. The verdict is now the sentinel's bottom crossing the sticky line, strict — a tolerance in either direction was itself a bug (late flashes transparent on pin, early paints it stuck at rest inside a scroll panel).

Also added `toolbarStuck.e2e.ts`, real-Chrome coverage for what the mocked specs can't reach: sub-pixel creep across the pin point, jump into pin range, pinning inside an overflow panel, below-the-fold. The sub-pixel case fails against every previous version on this branch.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 7]
```diff
@@ -133,10 +137,90 @@ export class LLMTestRunner {
   @Watch('stickyOffset')
   stickyOffsetChanged(newVal: number) {
     this.el.style.setProperty('--llmtr-sticky-top', `${newVal}px`);
+    this.updateToolbarStuck();
   }
 
   componentDidLoad() {
     this.el.style.setProperty('--llmtr-sticky-top', `${this.stickyOffset ?? 0}px`);
+    this.startToolbarStuckTracking();
+  }
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx:146]: Good catch — the guard was swallowing the first measurement. Fixed in 5c245e3 via componentDidRender, which runs after the first render. Added a spec that asserts the state is right after mount with no scroll dispatched.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 8]
```diff
@@ -133,10 +137,90 @@ export class LLMTestRunner {
...
+      return;
+    }
+    this.isTrackingToolbarStuck = true;
+    window.addEventListener('scroll', this.scheduleToolbarStuckUpdate, {
+      capture: true,
+      passive: true,
+    });
+    window.addEventListener('resize', this.scheduleToolbarStuckUpdate, {
+      passive: true,
+    });
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx:175]: Agreed. Same componentDidRender hook covers this — the error message appearing above the toolbar now re-runs the check. Shifts from host-page content above the component still aren't covered; leaving that unless it comes up.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 9]
```diff
@@ -133,10 +137,90 @@ export class LLMTestRunner {
...
+      this.updateToolbarStuck();
+    });
+  };
+
+  private updateToolbarStuck() {
+    if (!this.toolbarSentinelEl) {
+      return;
+    }
+    const scrollRoot = this.findScrollRoot();
+    const rootTop = scrollRoot ? scrollRoot.getBoundingClientRect().top : 0;
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx]: Right, the scrollport starts inside the border. Added clientTop to the root's rect in 5c245e3.

File: src/components/llm-test-runner/tests/toolbarStuck.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+async function isStuck(page: Awaited<ReturnType<typeof newE2EPage>>) {
+  return page.evaluate((cls: string) => {
+    const toolbar = document
+      .querySelector('llm-test-runner')
+      .shadowRoot.querySelector('.test-cases-toolbar');
+    return toolbar.classList.contains(cls);
+  }, STUCK_CLASS);
+}
+
+// IntersectionObserver delivers asynchronously, after layout.
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/tests/toolbarStuck.e2e.ts]: Yep, left over from the observer version. Comment now describes the rAF/render settle. Fixed in 5c245e3.

File: src/components/llm-test-runner/llm-test-runner.tsx [hunk 11]
```diff
@@ -133,10 +137,99 @@ export class LLMTestRunner {
...
   componentDidLoad() {
     this.el.style.setProperty('--llmtr-sticky-top', `${this.stickyOffset ?? 0}px`);
+    this.startToolbarStuckTracking();
+  }
+
+  // Catches shifts that move the sentinel without scrolling, such as the
+  // error message above the toolbar appearing.
+  componentDidRender() {
+    this.scheduleToolbarStuckUpdate();
+  }
```
[the remark this answers is not in the document]
Comment (ajay-fluxon) [reply src/components/llm-test-runner/llm-test-runner.tsx:152]: Guarded in 1d0fbb0 — put the check inside scheduleToolbarStuckUpdate rather than at the call site, so it covers the scroll and resize listeners too instead of just this one caller.

Worth noting the premise is narrower than it looks here: this package has no dist-hydrate-script output target and no ./hydrate export, and mock-doc does provide requestAnimationFrame, so there's no build where this actually threw. Took it anyway since it's cheap and matches the typeof window guard already in startToolbarStuckTracking.

**Review by arup-fluxon — APPROVED** [review]

**Comment by ajay-fluxon** [comment]
Closing in favor of #99. Rather than tracking a scroll-driven "stuck" state (sentinel + IntersectionObserver/scroll listener), the toolbar is now simply always opaque — no JS, no scroll tracking, no edge cases like the overflow-x: hidden on body one that came up here. Same underlying bleed-through bug, simpler fix.
